File: apps/api/src/services/jobPostingFetch.client.ts
import dnsPromises from "node:dns/promises";
import type { LookupAddress } from "node:dns";
import * as cheerio from "cheerio";
import { logger } from "../config/logger.js";
import { ValidationError } from "../errors/AppError.js";

/**
 * SSRF guard: the job URL is fully user-controlled and fetched from this
 * server, so without this check a user could point it at
 * http://localhost:4000/... or a cloud metadata endpoint
 * (169.254.169.254) and have the server fetch internal resources on their
 * behalf. Ported from HireConnect's job_posting_client.py.
 */
export class UnsafeJobUrlError extends ValidationError {}

const REQUEST_HEADERS = { "User-Agent": "Mozilla/5.0 (compatible; BelieveAI/1.0)" };
const TIMEOUT_MS = 15_000;
const ALLOWED_SCHEMES = new Set(["http:", "https:"]);
const MAX_REDIRECTS = 5;
// Below this, a fetch is treated as "didn't actually get the posting" rather than "the
// posting happens to be short" — well above a bare <title> shell page, well below a real posting.
const MIN_SUBSTANTIAL_CHARS = 200;

const WORKDAY_HOST_RE = /^([a-z0-9-]+)\.wd\d+\.myworkdayjobs\.com$/i;

function isPublicIpv4(ip: string): boolean {
  const parts = ip.split(".").map(Number);
  if (parts.length !== 4 || parts.some((p) => Number.isNaN(p))) return false;
  const [a, b] = parts as [number, number, number, number];
  if (a === 10) return false; // 10.0.0.0/8
  if (a === 127) return false; // loopback
  if (a === 0) return false; // "this network"
  if (a === 100 && b >= 64 && b <= 127) return false; // CGNAT 100.64.0.0/10
  if (a === 169 && b === 254) return false; // link-local, incl. cloud metadata
  if (a === 172 && b >= 16 && b <= 31) return false; // 172.16.0.0/12
  if (a === 192 && b === 168) return false; // 192.168.0.0/16
  if (a >= 224) return false; // multicast + reserved
  return true;
}

function isPublicIpv6(ip: string): boolean {
  const lower = ip.toLowerCase();
  if (lower === "::1") return false; // loopback
  if (lower.startsWith("fe80:") || lower.startsWith("fe8") || lower.startsWith("fe9") || lower.startsWith("fea") || lower.startsWith("feb")) return false; // link-local fe80::/10
  if (lower.startsWith("fc") || lower.startsWith("fd")) return false; // unique local fc00::/7
  if (lower.startsWith("::ffff:")) return isPublicIpv4(lower.slice(7)); // IPv4-mapped
  return true;
}

async function assertFetchableUrl(url: string): Promise<void> {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new UnsafeJobUrlError("That doesn't look like a valid link. Please paste the full job posting URL.");
  }

  if (!ALLOWED_SCHEMES.has(parsed.protocol)) {
    throw new UnsafeJobUrlError("Job posting links must start with http:// or https://. Please paste the full URL.");
  }
  if (!parsed.hostname) {
    throw new UnsafeJobUrlError("That link is missing a website address. Please paste the full job posting URL.");
  }

  let addresses: LookupAddress[];
  try {
    addresses = await dnsPromises.lookup(parsed.hostname, { all: true });
  } catch {
    throw new UnsafeJobUrlError(
      "We couldn't find that website. Check the link for typos, or paste the job description text instead.",
    );
  }

  for (const { address, family } of addresses) {
    const isPublic = family === 4 ? isPublicIpv4(address) : isPublicIpv6(address);
    if (!isPublic) {
      logger.warn({ host: parsed.hostname, address }, "jobPostingFetch: blocked non-public host");
      throw new UnsafeJobUrlError("That link points to a private address, so we can't open it.");
    }
  }
}

/** requests follows redirects itself in most clients, but that would bypass the SSRF check —
 * redirects are followed manually here, re-validating before each hop. */
async function fetchWithRedirectGuard(url: string, headers: Record<string, string>): Promise<Response> {
  let current = url;
  for (let i = 0; i < MAX_REDIRECTS; i++) {
    await assertFetchableUrl(current);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    let res: Response;
    try {
      res = await fetch(current, { headers, redirect: "manual", signal: controller.signal });
    } finally {
      clearTimeout(timer);
    }
    if (res.status < 300 || res.status >= 400) return res;
    const location = res.headers.get("location");
    if (!location) return res;
    current = new URL(location, current).toString();
  }
  throw new UnsafeJobUrlError("That link redirected too many times. Try the posting's direct URL.");
}

function isSubstantial(text: string): boolean {
  return text.length >= MIN_SUBSTANTIAL_CHARS;
}

function htmlFragmentToText(html: string | null | undefined): string {
  if (!html) return "";
  const $ = cheerio.load(html);
  $("script, style, noscript").remove();
  return $.root().text().replace(/\s+/g, " ").trim();
}

/** Flattens a whole career page, keeping the <title> as its own first line — the single
 * most reliable role/company signal on the page ("<Role> at <Company>", "<Role> | <Company>"). */
function pageHtmlToText(html: string): string {
  const $ = cheerio.load(html);
  const title = ($("title").first().text() || "").replace(/\s+/g, " ").trim();
  $("script, style, noscript, nav, header, footer, form, aside, svg").remove();
  const main = $("main").first().length ? $("main").first() : $('[role="main"]').first().length ? $('[role="main"]').first() : $("body");
  const bodyText = main.text().replace(/\s+/g, " ").trim();
  return [title, bodyText].filter(Boolean).join("\n");
}

async function fetchWorkday(scheme: string, host: string, path: string, tenant: string): Promise<string> {
  const apiUrl = `${scheme}//${host}/wday/cxs/${tenant}${path}`;
  try {
    const res = await fetchWithRedirectGuard(apiUrl, { ...REQUEST_HEADERS, Accept: "application/json" });
    if (!res.ok) return "";
    const data = (await res.json()) as { jobPostingInfo?: { title?: string; location?: string; locationsText?: string; jobDescription?: string } };
    const info = data.jobPostingInfo ?? {};
    const title = info.title ?? "";
    const location = info.location ?? info.locationsText ?? "";
    const description = htmlFragmentToText(info.jobDescription);
    return [title, location, description].filter(Boolean).join("\n").trim();
  } catch (err) {
    if (err instanceof UnsafeJobUrlError) throw err;
    return "";
  }
}

async function fetchGenericHtml(jobUrl: string): Promise<string> {
  try {
    const res = await fetchWithRedirectGuard(jobUrl, REQUEST_HEADERS);
    if (!res.ok) {
      logger.info({ jobUrl, status: res.status }, "jobPostingFetch: HTTP fetch failed");
      return "";
    }
    return pageHtmlToText(await res.text());
  } catch (err) {
    if (err instanceof UnsafeJobUrlError) throw err;
    logger.info({ jobUrl, err }, "jobPostingFetch: HTTP fetch errored");
    return "";
  }
}

/**
 * Fetches the raw text of a job posting. Two paths, tried in order (a
 * headless-browser third tier exists in the reference implementation for
 * fully client-rendered pages, deliberately not ported here — it needs a
 * bundled Chromium install, which is a meaningfully heavier deployment
 * footprint than this feature currently warrants):
 * 1. Workday-hosted postings — calls the same public JSON API the page's
 *    own frontend uses to render.
 * 2. Plain HTTP GET + HTML text extraction — works for most non-SPA career
 *    pages (Greenhouse, Lever, most company career pages).
 */
export async function fetchJobPostingText(jobUrl: string): Promise<string> {
  await assertFetchableUrl(jobUrl);
  const parsed = new URL(jobUrl);

  let bestSoFar = "";

  const workdayMatch = WORKDAY_HOST_RE.exec(parsed.hostname);
  if (workdayMatch) {
    const text = await fetchWorkday(parsed.protocol, parsed.host, parsed.pathname, workdayMatch[1]!);
    if (isSubstantial(text)) return text;
    if (text.length > bestSoFar.length) bestSoFar = text;
  }

  const text = await fetchGenericHtml(jobUrl);
  if (isSubstantial(text)) return text;
  if (text.length > bestSoFar.length) bestSoFar = text;

  return bestSoFar;
}

File: apps/backend/clients/job_posting_fetch_client.py
"""Mirrors apps/api's jobPostingFetch.client.ts.

SSRF guard: the job URL is fully user-controlled and fetched from this
server, so without this check a user could point it at
http://localhost:8000/... or a cloud metadata endpoint
(169.254.169.254) and have the server fetch internal resources on their
behalf. Uses stdlib `ipaddress`'s `is_global` rather than hand-rolling each
private/loopback/link-local/reserved range — same threat model as Node's
manual checks, fewer places to get a range wrong.
"""

from __future__ import annotations

import asyncio
import re
import socket
from ipaddress import ip_address
from urllib.parse import urljoin, urlparse

import httpx
from bs4 import BeautifulSoup

from core.errors import ValidationError
from core.logging import get_logger

logger = get_logger(__name__)


class UnsafeJobUrlError(ValidationError):
    pass


_REQUEST_HEADERS = {"User-Agent": "Mozilla/5.0 (compatible; BelieveAI/1.0)"}
_TIMEOUT_SECONDS = 15.0
_ALLOWED_SCHEMES = {"http", "https"}
_MAX_REDIRECTS = 5
# Below this, a fetch is treated as "didn't actually get the posting" rather than "the
# posting happens to be short" — well above a bare <title> shell page, well below a real posting.
_MIN_SUBSTANTIAL_CHARS = 200

_WORKDAY_HOST_RE = re.compile(r"^([a-z0-9-]+)\.wd\d+\.myworkdayjobs\.com$", re.IGNORECASE)


async def _assert_fetchable_url(url: str) -> None:
    parsed = urlparse(url)
    if parsed.scheme not in _ALLOWED_SCHEMES:
        raise UnsafeJobUrlError("Job posting links must start with http:// or https://. Please paste the full URL.")
    if not parsed.hostname:
        raise UnsafeJobUrlError("That link is missing a website address. Please paste the full job posting URL.")

    try:
        infos = await asyncio.to_thread(socket.getaddrinfo, parsed.hostname, None)
    except OSError as err:
        raise UnsafeJobUrlError(
            "We couldn't find that website. Check the link for typos, or paste the job description text instead."
        ) from err

    for info in infos:
        address = info[4][0]
        if not ip_address(address).is_global:
            logger.warning("jobPostingFetch: blocked non-public host %s (%s)", parsed.hostname, address)
            raise UnsafeJobUrlError("That link points to a private address, so we can't open it.")


async def _fetch_with_redirect_guard(url: str, headers: dict[str, str]) -> httpx.Response:
    """httpx follows redirects itself if asked, but that would bypass the SSRF check —
    redirects are followed manually here, re-validating before each hop."""
    current = url
    async with httpx.AsyncClient(follow_redirects=False, timeout=_TIMEOUT_SECONDS) as client:
        for _ in range(_MAX_REDIRECTS):
            await _assert_fetchable_url(current)
            res = await client.get(current, headers=headers)
            if res.status_code < 300 or res.status_code >= 400:
                return res
            location = res.headers.get("location")
            if not location:
                return res
            current = urljoin(current, location)
    raise UnsafeJobUrlError("That link redirected too many times. Try the posting's direct URL.")


def _is_substantial(text: str) -> bool:
    return len(text) >= _MIN_SUBSTANTIAL_CHARS


def _html_fragment_to_text(html: str | None) -> str:
    if not html:
        return ""
    soup = BeautifulSoup(html, "html.parser")
    for tag in soup(["script", "style", "noscript"]):
        tag.decompose()
    return re.sub(r"\s+", " ", soup.get_text()).strip()


def _page_html_to_text(html: str) -> str:
    """Flattens a whole career page, keeping the <title> as its own first line — the single
    most reliable role/company signal on the page ("<Role> at <Company>", "<Role> | <Company>")."""
    soup = BeautifulSoup(html, "html.parser")
    title_tag = soup.find("title")
    title = re.sub(r"\s+", " ", title_tag.get_text()).strip() if title_tag else ""
    for tag in soup(["script", "style", "noscript", "nav", "header", "footer", "form", "aside", "svg"]):
        tag.decompose()
    main = soup.find("main") or soup.find(attrs={"role": "main"}) or soup.find("body")
    body_text = re.sub(r"\s+", " ", main.get_text()).strip() if main else ""
    return "\n".join(filter(None, [title, body_text]))


async def _fetch_workday(scheme: str, host: str, path: str, tenant: str) -> str:
    api_url = f"{scheme}://{host}/wday/cxs/{tenant}{path}"
    try:
        res = await _fetch_with_redirect_guard(api_url, {**_REQUEST_HEADERS, "Accept": "application/json"})
        if res.status_code >= 400:
            return ""
        data = res.json()
        info = data.get("jobPostingInfo") or {}
        title = info.get("title") or ""
        location = info.get("location") or info.get("locationsText") or ""
        description = _html_fragment_to_text(info.get("jobDescription"))
        return "\n".join(filter(None, [title, location, description])).strip()
    except UnsafeJobUrlError:
        raise
    except Exception:  # noqa: BLE001 — best-effort probe, generic HTML fetch is the fallback
        return ""


async def _fetch_generic_html(job_url: str) -> str:
    try:
        res = await _fetch_with_redirect_guard(job_url, _REQUEST_HEADERS)
        if res.status_code >= 400:
            logger.info("jobPostingFetch: HTTP fetch failed for %s (status=%s)", job_url, res.status_code)
            return ""
        return _page_html_to_text(res.text)
    except UnsafeJobUrlError:
        raise
    except Exception as err:  # noqa: BLE001 — best-effort fetch, caller handles an empty result
        logger.info("jobPostingFetch: HTTP fetch errored for %s: %s", job_url, err)
        return ""


async def fetch_job_posting_text(job_url: str) -> str:
    """Fetches the raw text of a job posting. Two paths, tried in order:
    1. Workday-hosted postings — calls the same public JSON API the page's own frontend uses.
    2. Plain HTTP GET + HTML text extraction — works for most non-SPA career pages."""
    await _assert_fetchable_url(job_url)
    parsed = urlparse(job_url)

    best_so_far = ""

    workday_match = _WORKDAY_HOST_RE.match(parsed.hostname or "")
    if workday_match:
        text = await _fetch_workday(parsed.scheme, parsed.netloc, parsed.path, workday_match.group(1))
        if _is_substantial(text):
            return text
        if len(text) > len(best_so_far):
            best_so_far = text

    text = await _fetch_generic_html(job_url)
    if _is_substantial(text):
        return text
    if len(text) > len(best_so_far):
        best_so_far = text

    return best_so_far
